### PR TITLE
add minimumKubeletVersion

### PR DESCRIPTION
**Pull request by haircommander**

### CODE REVIEW

File: kubecontrolplane/v1/types.go [hunk 1]
```diff
@@ -62,6 +62,11 @@ type KubeAPIServerConfig struct {
 
 	// TODO this needs to be removed.
 	APIServerArguments map[string]Arguments `json:"apiServerArguments"`
+
+	// MinimumKubeletVersion is the lowest version of a kubelet that can meaningfully join the cluster.
```
Comment (sjenning) [inline kubecontrolplane/v1/types.go:66]: `s/can meaningfully join/is allowed to join/`

Comment (haircommander) [reply kubecontrolplane/v1/types.go:66]: it's allowed to join but it's blocked from doing anything meaningful in the cluster.

File: features/features.go [hunk 2]
```diff
@@ -662,4 +662,12 @@ var (
 								enhancementPR(legacyFeatureGateWithoutEnhancement).
 								enableIn(configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
 								mustRegister()
+
+	FeatureGateMinimumKubeletVersion = newFeatureGate("MinimumKubeletVersion").
+						reportProblemsToJiraComponent("Node").
+						contactPerson("haircommander").
+						productScope(ocpSpecific).
+						enableIn(configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
+						enhancementPR("https://github.com/kubernetes/enhancements/issues/4265").
```
Comment (deads2k) [inline features/features.go:671]: this points to ProcMount.  Can you update the link please?

File: config/v1/types_node.go [hunk 3]
```diff
@@ -46,6 +46,14 @@ type NodeSpec struct {
...
 	// +optional
 	WorkerLatencyProfile WorkerLatencyProfileType `json:"workerLatencyProfile,omitempty"`
+
+	// MinimumKubeletVersion is the lowest version of a kubelet that can meaningfully join the cluster.
+	// Specifically, the apiserver will deny (most) authorization requests of kubelets that are older
+	// than the specified version.
+	// +kubebuilder:validation:Pattern=`^[0-9]*\.[0-9]*\.[0-9]*$`
+	// +openshift:enable:FeatureGate=MinimumKubeletVersion
+	// +optional
+	MinimumKubeletVersion string `json:"minimumKubeletVersion,omitempty"`
```
Comment (enxebre) [inline config/v1/types_node.go:56]: do we really want to omitempty or keep the empty value in favour of discoverability?

File: config/v1/types_node.go [hunk 4]
```diff
@@ -46,6 +46,14 @@ type NodeSpec struct {
 	// the status and corresponding reaction of the cluster
 	// +optional
 	WorkerLatencyProfile WorkerLatencyProfileType `json:"workerLatencyProfile,omitempty"`
+
+	// MinimumKubeletVersion is the lowest version of a kubelet that can meaningfully join the cluster.
```
Comment (enxebre) [inline config/v1/types_node.go:50]: can we be more specific about what "meaningfully" entitle and about "deny (most)"?

File: kubecontrolplane/v1/types.go [hunk 5]
```diff
@@ -62,6 +62,12 @@ type KubeAPIServerConfig struct {
...
+	// Specifically, the apiserver will deny (most) authorization requests of kubelets that are older
```
Comment (enxebre) [inline kubecontrolplane/v1/types.go:67]: What's the expectation when this field and the one in v1Node contradict each other and how can an API consumer understand it?

File: config/v1/types_node.go [hunk 6]
```diff
@@ -46,6 +46,14 @@ type NodeSpec struct {
...
+	// Specifically, the apiserver will deny (most) authorization requests of kubelets that are older
+	// than the specified version.
+	// +kubebuilder:validation:Pattern=`^[0-9]*\.[0-9]*\.[0-9]*$`
+	// +openshift:enable:FeatureGate=MinimumKubeletVersion
```
Comment (enxebre) [inline config/v1/types_node.go:54]: Since this field is not marked as immutable, what happens if I change it to value higher than existing kubelets?

File: config/v1/types_node.go (hunk 3 above)
Comment (haircommander) [reply config/v1/types_node.go:56]: I think so, mostly because the other fields in this object also have it

File: config/v1/types_node.go (hunk 4 above)
Comment (haircommander) [reply config/v1/types_node.go:50]: yeah I added more details

File: config/v1/types_node.go (hunk 3 above)
Comment (enxebre) [reply config/v1/types_node.go:57]: cc @JoelSpeed can we define a criteria to be consistent here? Why would we add it here and decide not to here https://github.com/openshift/api/pull/2070?

File: config/v1/types_node.go (hunk 6 above)
Comment (enxebre) [reply config/v1/types_node.go:55]: ack. I think we should probably clarify somewhere that this validation on admission is by the nature of a distributed system only a best effort. Inflight operations might still result in older kubelet bypassing it if the right lucky timing take place.

For hypershift yes the norm is signal via conditions. For improving consumer UX we might also want to explore introducing VAP and let the current minimal kubelet be reported via [vap parameter](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#parameter-resources) that is inferred from e.g. the HC.status.

**Comment by enxebre** [comment]
@haircommander do we have a Jira/other way today to tag all related PRs for this effort in a single place? so it's easy for ourselves later in time or for others with less context to catch up?
E.g Add minimumKubeletVersion knob and validation epic involves
https://github.com/openshift/hypershift/pull/4980
https://github.com/openshift/kubernetes/pull/2104
https://github.com/openshift/api/pull/2059

File: config/v1/types_node.go (hunk 3 above)
Comment (JoelSpeed) [reply config/v1/types_node.go:57]: It may not be immediately obvious, but we touch on this in [our API conventions](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#configuration-vs-workload-apis).

This is a configuration API IMO and therefore should not have the `omitempty` tag. The section on pointers talks about this a little

File: config/v1/types_node.go [hunk 7]
```diff
@@ -46,6 +46,15 @@ type NodeSpec struct {
 	// the status and corresponding reaction of the cluster
 	// +optional
 	WorkerLatencyProfile WorkerLatencyProfileType `json:"workerLatencyProfile,omitempty"`
+
+	// MinimumKubeletVersion is the lowest version of a kubelet that can meaningfully join the cluster.
+	// Specifically, the apiserver will deny most authorization requests of kubelets that are older
+	// than the specified version, only allowing the kubelet to get and update its node object, and perform
+	// subjectaccessreviews.
+	// +kubebuilder:validation:Pattern=`^[0-9]*\.[0-9]*\.[0-9]*$`
```
Comment (JoelSpeed) [inline config/v1/types_node.go:54]: Prefer to use CEL for pattern matches now, as you can then provide a useful error message rather than just displaying the regex (and we believe users prefer prose to regex)

File: config/v1/types_node.go [hunk 8]
```diff
@@ -46,6 +46,15 @@ type NodeSpec struct {
...
+	// +openshift:enable:FeatureGate=MinimumKubeletVersion
+	// +optional
+	MinimumKubeletVersion string `json:"minimumKubeletVersion,omitempty"`
```
Comment (JoelSpeed) [inline config/v1/types_node.go:57]: What is the maximum length of this string? Should be added with a kubebuilder tag

**Comment by haircommander** [comment]
@enxebre there's also https://github.com/openshift/library-go/pull/1831 and https://github.com/openshift/cluster-kube-apiserver-operator/pull/1754

the overarching epic is https://issues.redhat.com/browse/OCPNODE-2506. Note a number of the cards tracking these PRs are closed because they were originally scoped to drafts for the work until I got feedback and could size the time to respond

File: config/v1/types_node.go (hunk 8 above)
Comment (haircommander) [reply config/v1/types_node.go:59]: updated!

File: config/v1/types_node.go (hunk 6 above)
Comment (JoelSpeed) [reply config/v1/types_node.go:57]: > but set a condition on the hosted cluster object, rather than block on admission

Why can't we block on admission?

File: config/v1/types_node.go [hunk 9]
```diff
@@ -46,6 +46,17 @@ type NodeSpec struct {
 	// the status and corresponding reaction of the cluster
 	// +optional
 	WorkerLatencyProfile WorkerLatencyProfileType `json:"workerLatencyProfile,omitempty"`
+
+	// MinimumKubeletVersion is the lowest version of a kubelet that can meaningfully join the cluster.
```
Comment (JoelSpeed) [inline config/v1/types_node.go:50]: I"d suggest to avoid subjective language like meaningfully, be direct like our product docs.
```suggestion
	// minimumKubeletVersion is the lowest version of a kubelet that can join the cluster.
```

File: config/v1/types_node.go [hunk 10]
```diff
@@ -46,6 +46,17 @@ type NodeSpec struct {
...
+	// Specifically, the apiserver will deny most authorization requests of kubelets that are older
+	// than the specified version, only allowing the kubelet to get and update its node object, and perform
+	// subjectaccessreviews.
```
Comment (JoelSpeed) [inline config/v1/types_node.go:53]: This is interesting technical detail, but, what does this mean to an end user?

File: config/v1/types_node.go [hunk 11]
```diff
@@ -46,6 +46,17 @@ type NodeSpec struct {
...
+	// Its max length is 8, so maximum version allowed is either "9.999.99" or "99.99.99".
```
Comment (JoelSpeed) [inline config/v1/types_node.go:54]: How was the maximum length decided? Is this a kubelet validation that we are transferring?

File: config/v1/tests/nodes.config.openshift.io/MinimumKubeletVersion.yaml [hunk 12]
```diff
@@ -0,0 +1,14 @@
+apiVersion: apiextensions.k8s.io/v1 # Hack because controller-gen complains if we don't have this
+name: "Node"
+crdName: nodes.config.openshift.io
+tests:
+  onCreate:
```
Comment (JoelSpeed) [inline config/v1/tests/nodes.config.openshift.io/MinimumKubeletVersion.yaml:5]: Should add some tests to test the validations you've added.

For example, making sure that the empty string is accepted, going too long, invalid formats etc

File: config/v1/types_node.go [hunk 13]
```diff
@@ -46,6 +46,17 @@ type NodeSpec struct {
...
+	// +kubebuilder:validation:XValidation:rule="self.matches('^[0-9]*.[0-9]*.[0-9]*$')",message="minmumKubeletVersion must be in a semver compatible format of x.y.z"
```
Comment (JoelSpeed) [inline config/v1/types_node.go:55]: I suspect you might need `self == \"\" || ` as a prefix to this rule
```suggestion
	// +kubebuilder:validation:XValidation:rule="self == \"\" || self.matches('^[0-9]*.[0-9]*.[0-9]*$')",message="minmumKubeletVersion must be in a semver compatible format of x.y.z"
```

File: kubecontrolplane/v1/types.go [hunk 14]
```diff
@@ -62,6 +62,13 @@ type KubeAPIServerConfig struct {
 
 	// TODO this needs to be removed.
 	APIServerArguments map[string]Arguments `json:"apiServerArguments"`
+
+	// MinimumKubeletVersion is the lowest version of a kubelet that can meaningfully join the cluster.
+	// Specifically, the apiserver will deny most authorization requests of kubelets that are older
+	// than the specified version, only allowing the kubelet to get and update its node object, and perform
+	// subjectaccessreviews.
+	// +openshift:enable:FeatureGate=MinimumKubeletVersion
+	MinimumKubeletVersion string `json:"minimumKubeletVersion,omitempty"`
```
Comment (JoelSpeed) [inline kubecontrolplane/v1/types.go:71]: Once we agree on godoc and validations on the other version of this field, please make sure to copy it here

File: config/v1/types_node.go [hunk 15]
```diff
@@ -46,6 +46,17 @@ type NodeSpec struct {
...
+	// +kubebuilder:validation:MaxLength:=8
+	// +openshift:enable:FeatureGate=MinimumKubeletVersion
+	// +optional
+	MinimumKubeletVersion string `json:"minimumKubeletVersion"`
```
Comment (JoelSpeed) [inline config/v1/types_node.go:59]: Is this Kubernetes versioning, or OCP versioning?

What would happen if the version string in the payload had a suffix, `1.2.3-ec.0`?

File: config/v1/types_node.go (hunk 10 above)
Comment (haircommander) [reply config/v1/types_node.go:53]: good question, I added it from @enxebre 's review, I'm not feeling too opinionated on keeping it

File: config/v1/types_node.go (hunk 11 above)
Comment (haircommander) [reply config/v1/types_node.go:54]: I decided semi arbitrarily based on what felt like a reasonable maximum. honestly, I don't think there's the notion of maximum length of kubernetes version, but practically I'd be surprised if this API lives long enough to hit this limit (10 years has gotten us to 1.31.0)

File: config/v1/types_node.go (hunk 10 above)
Comment (JoelSpeed) [reply config/v1/types_node.go:53]: I think Alberto's question was fair, I'm just not sure we are adding the right context

What does it look like to an end user when the kubelet is having these various authzs fail? 

Does it mean pods won't schedule there? Does it mean the node goes not ready? Is the node actually still functional?

File: config/v1/types_node.go (hunk 15 above)
Comment (haircommander) [reply config/v1/types_node.go:59]: kubernetes, and it would fail (despite technically being valid semver). I thought about allowing it but it doesn't feel useful to me personally

File: config/v1/types_node.go (hunk 10 above)
Comment (haircommander) [reply config/v1/types_node.go:53]: how does 
```
	// minimumKubeletVersion is the lowest version of a kubelet that can join the cluster.
	// Specifically, the apiserver will deny most authorization requests of kubelets that are older
	// than the specified version, only allowing the kubelet to get and update its node object, and perform
	// subjectaccessreviews.
	// This means the kubelet won't be able to view API objects it's responsible for running,
	// and will eventually be marked as NotReady.

``` 
work for you?

File: kubecontrolplane/v1/types.go (hunk 14 above)
Comment (haircommander) [reply kubecontrolplane/v1/types.go:71]: I decided to skip the validations because it didn't look like the other fields in this file were being validated. should I add them?

File: config/v1/types_node.go (hunk 15 above)
Comment (JoelSpeed) [reply config/v1/types_node.go:61]: Should the godoc specify that this is the Kubernetes version, specifically? So it's not confused later with OCP version?

Have you checked with the workloads folks that the reported K8s version is always going to be without a suffix? I'm wondering if that's true during a rebase and whether restricting versions with suffices would break the rebase somehow?

File: kubecontrolplane/v1/types.go (hunk 14 above)
Comment (JoelSpeed) [reply kubecontrolplane/v1/types.go:73]: Please. Over time we want to improve validation and even if validations aren't present for existing fields, we shouldn't use that as a pattern. Future tooling will enforce a lot of validation and leaving this now, will produce exceptions and toil in the future

File: config/v1/types_node.go (hunk 15 above)
Comment (haircommander) [reply config/v1/types_node.go:61]: so the version of the kubelet can have a suffix, as [it'll be stripped](https://github.com/openshift/kubernetes/pull/2104/files#diff-b7bdcb41a84b0a21cc83959481df9b457492b2e10b6d1350dff1278c8489bf44R91-R92) when comparing. The only thing that can't is the API. If a cluster needs to set the minimum version to an RC, they can set it to 1.x.0, and that'll include ecs

File: config/v1/types_node.go (hunk 10 above)
Comment (haircommander) [reply config/v1/types_node.go:53]: that should be blocked at admission: https://github.com/openshift/kubernetes/pull/2104/files#diff-5c437405eeba0789e9c42802e3f36cf6bdafd59d6d5c5dbaa6b66a2e02948bd7R117-R122 
So really this will only affect new kubelets that attempt to join the cluster after the min version is in place. those kubelets won't be able to run anything

**Comment by haircommander** [comment]
/retest

@JoelSpeed @enxebre any other thoughts?

File: config/v1/types_node.go (hunk 15 above)
Comment (JoelSpeed) [reply config/v1/types_node.go:65]: Is that worth capturing in the godoc for this field?

File: kubecontrolplane/v1/types.go (hunk 14 above)
Comment (JoelSpeed) [reply kubecontrolplane/v1/types.go:80]: You know that, and I now know that, but will future people who decide to copy this API? I realise it does nothing, but I'd rather we have it here for completeness, thanks for updating

File: config/v1/types_node.go (hunk 10 above)
Comment (JoelSpeed) [reply config/v1/types_node.go:53]: So perhaps then we take what you had above with a slight tweak?
```
// This means, any kubelet that attempts to join the cluster will not be able to run any assigned workloads, and will eventually be marked as not ready.
```

When you say will eventually be marked not ready, surely it will never get ready, since the CNI won't be able to initialise the network right?

File: config/v1/types_node.go (hunk 15 above)
Comment (haircommander) [reply config/v1/types_node.go:65]: good idea, I'll add a note

Comment (haircommander) [reply config/v1/types_node.go:67]: updated!

File: config/v1/types_node.go (hunk 10 above)
Comment (haircommander) [reply config/v1/types_node.go:53]: yeah unless it manages to upgrade itself which would require manual intervetion

Comment (haircommander) [reply config/v1/types_node.go:53]: updated to your suggestion

File: config/v1/tests/nodes.config.openshift.io/MinimumKubeletVersion.yaml [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+      expected: |
+        apiVersion: config.openshift.io/v1
+        kind: Node
+        spec: {}
+    - name: Should be able to create a minimumKubeletVersion
+      initial: |
+        apiVersion: config.openshift.io/v1
+        kind: Node
+        spec:
+          minimumKubeletVersion: 1.30.0
```
Comment (JoelSpeed) [inline config/v1/tests/nodes.config.openshift.io/MinimumKubeletVersion.yaml:21]: Please add a test that shows the empty string passes validation

**Review by JoelSpeed — COMMENTED** [review]
Once we have this final test case, I'm good to merge, thank you

File: config/v1/tests/nodes.config.openshift.io/MinimumKubeletVersion.yaml (hunk 16 above)
Comment (haircommander) [reply config/v1/tests/nodes.config.openshift.io/MinimumKubeletVersion.yaml:32]: done!

**Comment by JoelSpeed** [comment]
/lgtm

Thanks @haircommander
